src/raglite/_rag.py

```diff
@@ -95,25 +96,19 @@ def rag(  # noqa: PLR0913
     """Retrieval-augmented generation."""
     # Get the contexts for RAG as contiguous segments of chunks.
     config = config or RAGLiteConfig()
-    segments = _contexts(
+    segments = context_segments(
```

Comment:
I think it would be even better if `rag` and `async_rag` could accept `list[ContextSegment]`. That way, the developer is in full control of the input sources that RAG uses, and no additional querying is done on the developer's behalf by `rag`/`async_rag`.

Reply:
I agree, I will add it, but I think that these two methods are too overloaded. I think we should rethink the api surface. 

I am more inclined to separate it in 2 phases: retrieval and generation, but I don't want to do it now for the sake of back compatibility.

Reply:
We’re still pre-1.0, I’m not against making minor breaking changes if we can improve the API at this stage.

I think we should support fewer code paths here and simplify to only supporting _either_ a `SearchMethod` _or_ a `list[ContextSegment]`. Would you agree with that?

Reply:
I like the idea of being able to put the different steps in chainlit, so if I had to choose one I would choose the second, `list[ContextSegment]`. BTW, I start to dislike the name `ContextSegment`, I think `DocumentSegment` could be more appropriate. 

I am inclined to have 3 methods, not sure about the names right now. 

- Retrieve DocumentSegments 
- Rerank
- Call the LLM 

I think it is a good balance of separation of concerns, observability and low API surface. In any case, I would create another PR for that change.

Reply:
I am realising that reranking, currently only works at the level of chunks. So it should be:
- Retrieve chunks
- Rerank chunks
- combine chunks into `list[ContextSegment]`
- call the LLM

I am not thrilled of this solution, as it is weird for developers to have too many steps. An alternative solution would be to rerank the ContextSegments, but I have also my reserves about it.

Reply:
Yes that should be the order in my opinion too!

Reply:
I see how it seems like many steps, but at the same time I feel like each one is necessary. And four steps is still manageable I would say.